File: javascript/valid_parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var PARENTHESES = {
    '(': ')',
    '{': '}',
    '[': ']'
};

var isValid = function(s) {
    var stack = []
    var current_paren;
    
    for (var i = 0; i < s.length; i++) {
        current_paren = s[i];
        
        if (PARENTHESES[current_paren]) {
            stack.push(current_paren);
        } else if (current_paren !== PARENTHESES[stack.pop()]) {
            return false;
        }
    }
    
    return stack.length === 0;
};

File: ruby/valid_parentheses.rb
# Linked List Node implementation
class Node
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

# Stack Implementation
class Stack
  attr_accessor :head

  def push(val)
    new_node = Node.new(val)
    new_node.next = head
    self.head = new_node
    val
  end

  def pop
    if head
      popped_node = head
      self.head = head.next
      popped_node.val
    end
  end

  def peek
    head.val if head
  end

  def empty?
    head.nil?
  end
end

PARENTHESES = { '(' => ')', '{' => '}', '[' => ']' }

def is_valid(s)
  stack = Stack.new
  parentheses_only = s.gsub(/[^\(\)\[\]\{\}]/, '')
  return false if parentheses_only.size <= 1

  parentheses_only.chars.each do |paren|
    if PARENTHESES[paren]
      stack.push(paren)
    else
      return false if paren != PARENTHESES[stack.pop]
    end
  end

  stack.empty?
end
